### PR TITLE
BUG: Check for null pointer before fitting volume

### PR DESCRIPTION
**Pull request by pieper**
See discussion here: https://discourse.slicer.org/t/slicer-preview-crashes-on-slicer-util-resetsliceviews/40488/4

### CODE REVIEW

**Review by lassoan — APPROVED** [review]
Thank you, it looks good to me. I've added some more checks and small additional improvements, please have a look @pieper.

**Comment by pieper** [comment]
> Thank you, it looks good to me. I've added some more checks and small additional improvements, please have a look @pieper.

I'm not allowed to review since it's my PR but LGTM 👍 

N

**Comment by pieper** [comment]
Build error looks unrelated.

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 1]
```diff
@@ -2706,3 +2658,17 @@ bool vtkMRMLSliceLogic::GetSliceOffsetRangeResolution(double range[2], double& r
 
   return true;
 }
+
+//----------------------------------------------------------------------------
+vtkMRMLVolumeNode* vtkMRMLSliceLogic::GetFirstVolumeNode()
+{
+  for (int layer = 0; layer < Layer_Last; layer++)
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:2665]:
```suggestion
  for (int layer = 0; layer < Self::Layer_Last; layer++)
```

Then consider adding the following to the class declaration

```cpp
typedef vtkMRMLSliceLogic Self;
```


```diff
class VTK_MRML_LOGIC_EXPORT vtkMRMLSliceLogic : public vtkMRMLAbstractLogic
{
public:

+  typedef vtkMRMLSliceLogic Self;

  /// The Usual VTK class functions
  static vtkMRMLSliceLogic *New();
  vtkTypeMacro(vtkMRMLSliceLogic,vtkMRMLAbstractLogic);
  void PrintSelf(ostream& os, vtkIndent indent) override;

```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 2]
```diff
@@ -2418,16 +2382,12 @@ int vtkMRMLSliceLogic::GetSliceIndexFromOffset(double sliceOffset, vtkMRMLVolume
...
+  if (volumeNode)
   {
-    volumeNode = this->GetLayerVolumeNode (layer);
-    if (volumeNode)
-    {
-      int sliceIndex=this->GetSliceIndexFromOffset( sliceOffset, volumeNode );
-      // return the result for the first available layer
-      return sliceIndex;
-    }
+    int sliceIndex=this->GetSliceIndexFromOffset(sliceOffset, volumeNode);
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:2388]:
```suggestion
    int sliceIndex = this->GetSliceIndexFromOffset(sliceOffset, volumeNode);
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 3]
```diff
@@ -1742,34 +1743,30 @@ void vtkMRMLSliceLogic::FitSliceToVolumes(vtkCollection *volumeNodes, int width,
...
+  vtkMRMLVolumeNode *backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
   return (this->GetVolumeSliceSpacing(backgroundNode));
 }
 
 //----------------------------------------------------------------------------
 void vtkMRMLSliceLogic::GetBackgroundSliceBounds(double sliceBounds[6])
 {
-  vtkMRMLVolumeNode *backgroundNode = nullptr;
-  backgroundNode = this->GetLayerVolumeNode(0);
+  vtkMRMLVolumeNode *backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1769]:
```suggestion
  vtkMRMLVolumeNode* backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 4]
```diff
@@ -1843,7 +1834,7 @@ void vtkMRMLSliceLogic::FitSliceToAll(int width, int height)
   }
 
   vtkNew<vtkCollection> volumeNodes;
-  for (int layer = 0; layer < 3; layer++)
+  for (int layer = 0; layer < vtkMRMLSliceLogic::Layer_Last; layer++)
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1837]:
```suggestion
  for (int layer = 0; layer < Self::Layer_Last; layer++)
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 5]
```diff
@@ -1958,82 +1949,55 @@ void vtkMRMLSliceLogic::ResizeSliceNode(double newWidth, double newHeight)
...
-  for (i=0; i<3; i++)
-  {
-    sliceBounds[2*i]   = LARGE_BOUNDS_NUM;
-    sliceBounds[2*i+1] = SMALL_BOUNDS_NUM;
-  }
-
-  vtkMRMLVolumeNode *volumeNode;
-  for ( int layer=0; layer < 3; layer++ )
+  vtkBoundingBox sliceBoundingBox;
+  for (int layer=0; layer < vtkMRMLSliceLogic::Layer_Last; layer++)
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1977]:
```suggestion
  for (int layer = 0; layer < Self::Layer_Last; layer++)
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 6]
```diff
@@ -1958,82 +1949,55 @@ void vtkMRMLSliceLogic::ResizeSliceNode(double newWidth, double newHeight)
...
-    {
-      sliceBounds[2*i] = -100;
-    }
-    if (sliceBounds[2*i+1] == SMALL_BOUNDS_NUM)
+    sliceBoundingBox.GetBounds(sliceBounds);
+  }
+  else
+  {
+    // no volumes are shown, set some valid bounds to avoid singularities
+    for (int i = 0; i < 3; i++)
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1995]: `i` -> `dim` (or `idx`)

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 7]
```diff
@@ -2418,16 +2382,12 @@ int vtkMRMLSliceLogic::GetSliceIndexFromOffset(double sliceOffset, vtkMRMLVolume
 // sliceIndex: DICOM slice index, 1-based
 int vtkMRMLSliceLogic::GetSliceIndexFromOffset(double sliceOffset)
 {
-  vtkMRMLVolumeNode *volumeNode;
-  for (int layer=0; layer < 3; layer++ )
+  vtkMRMLVolumeNode* volumeNode = vtkMRMLSliceLogic::GetFirstVolumeNode();
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:2385]:
```suggestion
  vtkMRMLVolumeNode* volumeNode = vtkMRMLSliceLogic::GetFirstVolumeNode();
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx (hunk 5 above)
Comment (lassoan) [reply Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1977]: I would keep the VTK style of using the actual class name

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx (hunk 6 above)
Comment (lassoan) [reply Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1995]: ok!

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx (hunk 7 above)
Comment (lassoan) [reply Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:2385]: I don't see any change

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 8]
```diff
@@ -1742,34 +1743,30 @@ void vtkMRMLSliceLogic::FitSliceToVolumes(vtkCollection *volumeNodes, int width,
 // Get the size of the volume, transformed to RAS space
 void vtkMRMLSliceLogic::GetBackgroundRASBox(double rasDimensions[3], double rasCenter[3])
 {
-  vtkMRMLVolumeNode *backgroundNode = nullptr;
-  backgroundNode = this->GetLayerVolumeNode(0);
+  vtkMRMLVolumeNode *backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1746]:
```suggestion
  vtkMRMLVolumeNode* backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 9]
```diff
@@ -1742,34 +1743,30 @@ void vtkMRMLSliceLogic::FitSliceToVolumes(vtkCollection *volumeNodes, int width,
...
   this->GetVolumeRASBox(backgroundNode, rasDimensions, rasCenter);
 }
 
 //----------------------------------------------------------------------------
 // Get the size of the volume, transformed to RAS space
 void vtkMRMLSliceLogic::GetBackgroundSliceDimensions(double sliceDimensions[3], double sliceCenter[3])
 {
-  vtkMRMLVolumeNode *backgroundNode = nullptr;
-  backgroundNode = this->GetLayerVolumeNode(0);
+  vtkMRMLVolumeNode *backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1754]:
```suggestion
  vtkMRMLVolumeNode* backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 10]
```diff
@@ -1742,34 +1743,30 @@ void vtkMRMLSliceLogic::FitSliceToVolumes(vtkCollection *volumeNodes, int width,
...
   this->GetVolumeSliceDimensions(backgroundNode, sliceDimensions, sliceCenter);
 }
 
 //----------------------------------------------------------------------------
 // Get the spacing of the volume, transformed to slice space
 double *vtkMRMLSliceLogic::GetBackgroundSliceSpacing()
 {
-  vtkMRMLVolumeNode *backgroundNode = nullptr;
-  backgroundNode = this->GetLayerVolumeNode(0);
+  vtkMRMLVolumeNode *backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1762]:
```suggestion
  vtkMRMLVolumeNode* backgroundNode = this->GetLayerVolumeNode(vtkMRMLSliceLogic::LayerBackground);
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx (hunk 1 above)
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx]:
```suggestion
  for (int layer = 0; layer < vtkMRMLSliceLogic::Layer_Last; layer++)
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 11]
```diff
@@ -1491,10 +1488,10 @@ void vtkMRMLSliceLogic::GetVolumeRASBox(vtkMRMLVolumeNode *volumeNode, double ra
   double bounds[6];
   volumeNode->GetRASBounds(bounds);
 
-  for (int i=0; i<3; i++)
+  for (int axis =0; axis <3; axis++)
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1491]:
```suggestion
  for (int axis = 0; axis < 3; axis++)
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx (hunk 5 above)
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:1977]:
```suggestion
  for (int layer = 0; layer < vtkMRMLSliceLogic::Layer_Last; layer++)
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 12]
```diff
@@ -2285,9 +2249,9 @@ std::vector< vtkMRMLDisplayNode*> vtkMRMLSliceLogic::GetPolyDataDisplayNodes()
   std::vector<vtkMRMLSliceLayerLogic *> layerLogics;
   layerLogics.push_back(this->GetBackgroundLayer());
   layerLogics.push_back(this->GetForegroundLayer());
-  for (unsigned int i=0; i<layerLogics.size(); i++)
+  for (unsigned int layerIndex = 0; layerIndex<layerLogics.size(); layerIndex++)
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:2252]:
```suggestion
  for (unsigned int layerIndex = 0; layerIndex < layerLogics.size(); layerIndex++)
```

File: Libs/MRML/Logic/vtkMRMLSliceLogic.cxx [hunk 13]
```diff
@@ -2285,9 +2249,9 @@ std::vector< vtkMRMLDisplayNode*> vtkMRMLSliceLogic::GetPolyDataDisplayNodes()
...
   {
-    vtkMRMLSliceLayerLogic *layerLogic = layerLogics[i];
+    vtkMRMLSliceLayerLogic *layerLogic = layerLogics[layerIndex];
```
Comment (jcfr) [inline Libs/MRML/Logic/vtkMRMLSliceLogic.cxx:2254]:
```suggestion
    vtkMRMLSliceLayerLogic* layerLogic = layerLogics[layerIndex];
```

**Review by jcfr — APPROVED** [review]
Thanks for reviewing and integrating the suggestions :pray:

**Comment by jcfr** [comment]
How about squashing all "Update ..." commits into `ENH: Make vktMRMLSliceLogic class more robust` so that there are only two commits ?

**Comment by lassoan** [comment]
> How about squashing all "Update ..." commits into `ENH: Make vktMRMLSliceLogic class more robust` so that there are only two commits ?

That's exactly what I've been planning to do. I'm building now, will do a quick test, then push and merge.

**Review by lassoan — APPROVED** [review]

**Pull request merged**